holoviews/plotting/bokeh/callbacks.py

```diff
@@ -547,14 +566,177 @@ def _process_msg(self, msg):
+            margin: -0.5em;
+            outline: none;
+            box-shadow: none;
+            position: absolute;
+            top: 0;
+            right: 0;
+        }
+        .bk-btn::after {
+            content: '\2715';
+        }
+        """],
+        css_classes=["popup-close-btn"])
+        self._panel = Panel(
+            position=XY(x=np.nan, y=np.nan),
+            anchor="top_left",
+            elements=[close_button],
+            visible=False
+        )
+        close_button.js_on_click(CustomJS(args=dict(panel=self._panel), code="panel.visible = false"))
+
+        self.plot.state.elements.append(self._panel)
+        self._watch_position()
+
+    def _watch_position(self):
+        geom_type = self.geom_type
+        self.plot.state.on_event('selectiongeometry', self._update_selection_event)
+        self.plot.state.js_on_event('selectiongeometry', CustomJS(
+            args=dict(panel=self._panel),
+            code=f"""
+            export default ({{panel}}, cb_obj, _) => {{
+              const el = panel.elements[1]
+              if ((el && !el.visible) || !cb_obj.final || ({geom_type!r} !== 'any' && cb_obj.geometry.type !== {geom_type!r})) {{
+                 return
+              }}
+              let pos;
+              if (cb_obj.geometry.type === 'point') {{
+                pos = {{x: cb_obj.geometry.x, y: cb_obj.geometry.y}}
+              }} else if (cb_obj.geometry.type === 'rect') {{
+                pos = {{x: cb_obj.geometry.x1, y: cb_obj.geometry.y1}}
+              }} else if (cb_obj.geometry.type === 'poly') {{
+                pos = {{x: Math.max(...cb_obj.geometry.x), y: Math.max(...cb_obj.geometry.y)}}
+              }}
+              if (pos) {{
+                panel.position.setv(pos)
+              }}
+            }}""",
+        ))
+
+    def _get_position(self, event):
+        if self.geom_type not in ('any', event.geometry['type']):
+            return
+        elif event.geometry['type'] == 'point':
+            return dict(x=event.geometry['x'], y=event.geometry['y'])
+        elif event.geometry['type'] == 'rect':
+            return dict(x=event.geometry['x1'], y=event.geometry['y1'])
+        elif event.geometry['type'] == 'poly':
+            return dict(x=np.max(event.geometry['x']), y=np.max(event.geometry['y']))
+
+    def _update_selection_event(self, event):
+        self._selection_event = event
+
+    def on_msg(self, msg):
+        super().on_msg(msg)
+        event = self._selection_event
+        if event is not None:
+            if self.geom_type not in (event.geometry["type"], "any") or not event.final:
+                return
+
+        for stream in self.streams:
+            popup = stream.popup
+            if popup is not None:
+                break
+
+        if callable(popup):
+            popup = popup(**stream.contents)
+
+        # If no popup is defined, hide the panel
+        if popup is None:
+            if self._panel.visible:
```

Comment:
Just saw this error in an app:

``` python-traceback
2024-04-17 09:58:13,275 Exception in callback functools.partial(<bound method IOLoop._discard_future_result of <tornado.platform.asyncio.AsyncIOMainLoop object at 0x76878123fc50>>, <Task finished name='Task-155' coro=<ServerSession.with_document_locked() done, defined at /home/shh/miniconda3/envs/holoviz/lib/python3.12/site-packages/bokeh/server/session.py:77> exception=AttributeError("'BoundsCallback' object has no attribute '_panel'")>)
Traceback (most recent call last):
  File "/home/shh/miniconda3/envs/holoviz/lib/python3.12/site-packages/tornado/ioloop.py", line 750, in _run_callback
    ret = callback()
          ^^^^^^^^^^
  File "/home/shh/miniconda3/envs/holoviz/lib/python3.12/site-packages/tornado/ioloop.py", line 774, in _discard_future_result
    future.result()
  File "/home/shh/miniconda3/envs/holoviz/lib/python3.12/site-packages/bokeh/server/session.py", line 98, in _needs_document_lock_wrapper
    result = await result
             ^^^^^^^^^^^^
  File "/home/shh/projects/holoviz/repos/holoviews/holoviews/plotting/bokeh/callbacks.py", line 338, in on_event
    await self.process_on_event()
  File "/home/shh/projects/holoviz/repos/holoviews/holoviews/plotting/bokeh/callbacks.py", line 363, in process_on_event
    self.on_msg(msg)
  File "/home/shh/projects/holoviz/repos/holoviews/holoviews/plotting/bokeh/callbacks.py", line 677, in on_msg
    if self._panel.visible:
       ^^^^^^^^^^^
AttributeError: 'BoundsCallback' object has no attribute '_panel'

```

Reply:
Indeed, that needs a guard.

Reply:
Is this it?
https://github.com/holoviz/holoviews/pull/6168/files#diff-c0b814bedcee02b97de63fa13fc6a00858b3def2afc67eef7cdfaf04524be863R670